test/json/json_generator_test.rb

```diff
@@ -536,6 +536,72 @@ def test_to_json_called_with_state_object
     assert_instance_of JSON::State, argument
   end
 
+  module CustomToJSON
+    def to_json(*)
+      %{"#{self.class.name}#to_json"}
+    end
+  end
+
+  module CustomToS
+    def to_s
+      "#{self.class.name}#to_s"
+    end
+  end
+
+  class ArrayWithToJSON < Array
+    include CustomToJSON
+  end
+
+  def test_array_subclass_with_to_json
+    assert_equal '"JSONGeneratorTest::ArrayWithToJSON#to_json"', JSON.generate(ArrayWithToJSON.new)
+    assert_equal '{"[]":1}', JSON.generate(ArrayWithToJSON.new => 1)
+  end
+
+  class ArrayWithToS < Array
+    include CustomToS
+  end
+
+  def test_array_subclass_with_to_s
+    assert_equal '[]', JSON.generate(ArrayWithToS.new)
+    assert_equal '{"JSONGeneratorTest::ArrayWithToS#to_s":1}', JSON.generate(ArrayWithToS.new => 1)
```

Comment:
This doesn't seem reasonable, i.e. calling `to_s` on an Array subclass (both semantically and performance-wise).
Inheriting from core classes is frowned upon, so without a convincing use case I see no value to respect those old weird/broken semantics.

Reply:
I think ideally we wouldn't treat subclasses specially at all, it doesn't seem reasonable to subclass a core type and then expect different behavior than the superclass for JSON behavior and many other things (Liskov substitution principle).
If someone wants to have some custom serialization to JSON, it seems simple enough to have a custom type not inheriting from a core type and having `to_json` implemented on it.

Reply:
I'm not disagreeing, I'm just saying we should be very deliberate about it, because backward compatibility is very important.

Reply:
Ah for this specific example I missed that `to_s` is only called because the `ArrayWithToS` is a Hash key, that seems less crazy at least :sweat_smile: